teps/0010-optional-workspaces.md

```diff
@@ -0,0 +1,367 @@
+
+### Goals
+
+1. Allow Tasks to declare optional support for file-based credentials.
+2. Enable Tasks to offer support of Tekton's multiple competing credential mechanisms.
+3. Allow Tasks to optionally receive storage with some data to use during execution.
+4. Allow Tasks to optionally receive storage on to which the Task will place some output data.
+5. Enable Tasks to detect if any declared optional storage has been provided or not at runtime.
+
+### Non-Goals
+
+1. Declaring / binding specific file content on the optional storage.
+2. Validation of the file contents of the optional storage.
+3. Validation of "competing" filesystem fields used in a single TaskRun. An optional Workspace
+and an optional PipelineResource competing for the same location on the filesystem will be left
+up to the runtime to handle.
+
+## Requirements
+
+When optional storage is not provided by a TaskRun the Task must be able to
+detect this within the container / script.
+
+Tekton Variables should fail gracefully in the event that a workspace is
+not provided. Using a variable like `$(workspaces.foo.path)` should not
+cause the controller to abandon or fail a TaskRun when the optional workspace
+`foo` is omitted by it. Likewise the literal string `"$(workspaces.foo.path)"`
+should not end up being passed to the Task.
+
+When a Tekton deployment includes configuration for
+[Default TaskRun Workspaces](https://github.com/tektoncd/pipeline/blob/3a7a693429c605a67e30efa5b74675e403be79bb/config/config-defaults.yaml#L66-L70)
+a missing optional Workspace does not result in a Default TaskRun Workspace
+being injected in its place. The reason for this is that optional behaviour
+in a Task should not be switched on automatically by the Default TaskRun
+Workspace mechanism.
+
+## Proposal
+
+Allow Workspaces to be declared "optional" in Tasks. An optional Workspace is
+not required for successful execution of the Task but can be provided by a TaskRun
+to configure some optional behaviour, like using a specific kind of credential.
+
+An optional Workspace provides the following variables:
+
+- `$(workspaces.<workspace-name>.bound)`: the literal string `"true"` or `"false"`
+reflecting whether the optional workspace has been bound or not.
+- `$(workspaces.<workspace-name>.path)`: the path to the workspace if it has been
+bound _or_ an empty string if it has not been.
+
+A required Workspace today already provides a `path` variable. Required Workspaces
+will be updated to also expose the following variable:
+
+- `$(workspaces.<workspace-name>.bound)`: the literal string `"true"`. A required
+Workspace must always be bound by runtime.
+
+
+### User Stories
+
+#### Story 1
+
+As a Catalog Task author I want to publish a Task that can support both Workspaces
+and PipelineResources in the same location on a Step's filesystem so that my Task's
+users can choose whether to use either field without requiring separate Tasks.
```

Comment:
oh interesting! i hadnt thought about optional workspaces allowing this 🤔  would this work without changing PipelineResources? i think that if a task says it needs a PipelineResource, the Task won't run without it? (i.e. we'd have to make PipelineResources optional too?)

Reply:
@bobcatfish we do support optional PipelineResource already :wink:

Reply:
@vdemeester beat me to it!

Reply:
ooo TIL